### ISSUE TITLE
Network always predicts no mask

### ISSUE THREAD

**Issue by jbitton** [jakeret/tf_unet#234]
Hi there!

I've been trying to train a u-net using your repo on the kaggle ultrasound nerve dataset. However, no matter what I do, the mask I get is always all zeros. This is how I process the prediction:

```python
mask = np.zeros((prediction.shape[1], prediction.shape[2]), dtype=np.uint8)
for i in range(prediction.shape[1]):
    for j in range(prediction.shape[2]):
        if  prediction[0][i][j][1] >= 0.5:
            print('lit')
        mask[i][j] = np.argmax(prediction[0][i][j])
```

I've looked through previous issues, and I followed the code available in https://github.com/jakeret/tf_unet/issues/3#issuecomment-260112160 (for some reason, the version in the scripts folder has been throwing errors).

I think that it may always predict no mask is because the network achieves a high enough accuracy doing so. What am I doing wrong here? How can I actually get predictions?

Here are some logs from the end of training:
```
2018-12-15 23:26:12,042 Iter 3198, Minibatch Loss= 0.4196, Training Accuracy= 0.8665, Minibatch error= 13.3%
2018-12-15 23:26:14,717 Epoch 99, Average loss: 0.5016, learning rate: 0.0010
2018-12-15 23:26:14,791 Verification error= 37.4%, loss= 0.7187
2018-12-15 23:26:20,868 Optimization Finished!
..................
Using data from: /scratch/jtb470/nerve_split_data/train
Number of data_files used: 3060
Testing error rate: 0.00%
```

Thanks in advance.

**Comment by jakeret** [comment]
hi @jbitton there might be several reason that the model does not predict anything.
Very recently someone found a bug in the `_process_data ` function (#228). Maybe you encounter the same issue.
An other thing: if I remember correctly the nerve segmentation data set is highly imbalanced. Hence using average as loss is not very suited. Have you tried to use a different one (e.g. dice_coefficient)

**Comment by jbitton** [comment]
@jakeret hi, thanks for the response. I added the `if` statements to protect against division by zero errors and changed my loss function to the dice coefficient. I just restarted training and training accuracy / minibatch loss still seems strange:

```
2018-12-19 11:17:55,138 Verification error= 100.0%, loss= -0.4902
2018-12-19 11:17:57,601 Start optimization
2018-12-19 11:18:22,779 Iter 0, Minibatch Loss= -0.6293, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:18:35,344 Iter 1, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:18:48,181 Iter 2, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:00,780 Iter 3, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:13,417 Iter 4, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:26,560 Iter 5, Minibatch Loss= -0.6047, Training Accuracy= 0.6047, Minibatch error= 39.5%
2018-12-19 11:19:39,370 Iter 6, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:52,543 Iter 7, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:05,163 Iter 8, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:17,818 Iter 9, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:30,857 Iter 10, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:17:55,138 Verification error= 100.0%, loss= -0.4902
2018-12-19 11:17:57,601 Start optimization
2018-12-19 11:18:22,779 Iter 0, Minibatch Loss= -0.6293, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:18:35,344 Iter 1, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:18:48,181 Iter 2, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:00,780 Iter 3, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:13,417 Iter 4, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:26,560 Iter 5, Minibatch Loss= -0.6047, Training Accuracy= 0.6047, Minibatch error= 39.5%
2018-12-19 11:19:39,370 Iter 6, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:19:52,543 Iter 7, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:05,163 Iter 8, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:17,818 Iter 9, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
2018-12-19 11:20:30,857 Iter 10, Minibatch Loss= -1.0000, Training Accuracy= 1.0000, Minibatch error= 0.0%
```

I highly doubt the accuracy is so good. Anything else you can think of doing?

**Comment by rytisss** [comment]
I am in same situation....
![image](https://user-images.githubusercontent.com/22094617/52178917-90f87080-27dc-11e9-9db5-f6a4fc63344c.png)

**Comment by rytisss** [comment]
i dont know if dice_coeficient is the most suitable loss function when i have more than half of data as negatives, case it should give 0.0 lots of time. Btw i am training on DAGM dataset and i received very good results with cross_entropy loss function and i cant receive any results with dice_coeficient

**Comment by rytisss** [comment]
Tried to make loss range from 0 to 1, by writing loss = 1.0 - (2 * intersection / (union)), but still end up same

**Comment by rytisss** [comment]
learning_rate = 0.0001 or a bit bigger, optimizer = 'adam'
results:
![image](https://user-images.githubusercontent.com/22094617/52202039-fc425100-2875-11e9-99a6-fdcf363e29a2.png)
It trains pretty well

**Comment by jis478** [comment]
I think predicting all black is caused by the imbalanced nature of your training dataset. Can you try a modified version I've made based on this repo? I'm wondering my approach also works for other imbalanced problems.

You can simply try the below jupyter notebook file.
https://github.com/jis478/Tensorflow/tree/master/Unet_modified/example/membrane/code/Unet_modified_execution.ipynb
